File: DAA/C Implementation/Week5/1.c
#include <stdio.h>
#include <stdlib.h>

struct graph
{
    char src;
    char dest;
    int weight;
};

int comparator(const void *p1, const void *p2)
{
    const struct graph *x = p1;
    const struct graph *y = p2;
    return x->weight - y->weight;
}

void makeSet(int parent[], int rank[], int n)
{
    for (int i = 0; i < n; i++)
    {
        parent[i] = i;
        rank[i] = 0;
    }
}

int findParent(int parent[], int component)
{
    if (parent[component] == component)
        return component;

    return parent[component] = findParent(parent, parent[component]);
}

void unionSet(int u, int v, int parent[], int rank[], int n)
{
    u = findParent(parent, u);
    v = findParent(parent, v);
    if (rank[u] < rank[v])
        parent[u] = v;
    else if (rank[u] > rank[v])
        parent[v] = u;
    else
    {
        parent[v] = u;
        rank[u]++;
    }
}

void kruskalMST(int v, int n, struct graph g[])
{

    qsort(g, n, sizeof(g[0]), comparator);
    int parent[v];
    int rank[v];
    makeSet(parent, rank, n);

    int minCost = 0;

    printf("Edges \tWeight\n");
    for (int i = 0; i < n; i++)
    {
        char v1 = g[i].src;
        char v2 = g[i].dest;
        int wt = g[i].weight;
        int p1 = v1 - 'A';
        int p2 = v2 - 'A';
        if (findParent(parent, p1) != findParent(parent, p2))
        {
            unionSet(p1, p2, parent, rank, v);
            minCost += wt;
            printf("%c - %c \t%d\n", v1, v2, wt);
        }
    }
    printf("Minimum Spanning Tree: %d\n", minCost);
}

int main()
{
    printf("Kruskal's Algorithm in C\n");
    int vertices = 5;
    struct graph g[] =
        {
            {'A', 'B', 2},
            {'A', 'C', 7},
            {'A', 'D', 2},
            {'B', 'D', 5},
            {'B', 'E', 7},
            {'C', 'D', 5},
            {'D', 'E', 4}};
    int numEdges = sizeof(g) / sizeof(struct graph);
    kruskalMST(vertices, numEdges, g);

    return 0;
}

File: DAA/Week5/1.py
class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = []

    def addEdge(self, u, v, w):
        u_index = ord(u) - 65
        v_index = ord(v) - 65
        self.graph.append([u_index, v_index, w])

    def find(self, parent, i):
        if parent[i] != i:
            parent[i] = self.find(parent, parent[i])
        return parent[i]

    def union(self, parent, rank, x, y):
        if rank[x] < rank[y]:
            parent[x] = y
        elif rank[x] > rank[y]:
            parent[y] = x
        else:
            parent[y] = x
            rank[x] += 1

    def kruskalMST(self):
        result = []
        i = 0
        e = 0
        self.graph = sorted(self.graph, key=lambda item: item[2])
        parent = []
        rank = []
        for node in range(self.V):
            parent.append(node)
            rank.append(0)
        while e < self.V - 1:
            u, v, w = self.graph[i]
            i = i + 1
            x = self.find(parent, u)
            y = self.find(parent, v)
            if x != y:
                e = e + 1
                result.append([u, v, w])
                self.union(parent, rank, x, y)
        minimumCost = 0
        print("Edge \tWeight")
        for u, v, weight in result:
            minimumCost += weight
            print(f"{chr(u+65)} - {chr(v+65)} \t{weight}")
        print("Minimum Spanning Tree:", minimumCost)


if __name__ == '__main__':
    print("Kruskal's algorithm in Python")
    g = Graph(5)
    g.addEdge('A', 'B', 2)
    g.addEdge('A', 'C', 7)
    g.addEdge('A', 'D', 2)
    g.addEdge('B', 'D', 5)
    g.addEdge('B', 'E', 7)
    g.addEdge('C', 'D', 5)
    g.addEdge('D', 'E', 4)
    g.kruskalMST()